src/engine/client/dl_main.cpp

```diff
@@ -194,13 +194,48 @@ class FileDownload : public CurlDownload {
 	FileDownload(Str::StringRef url, FS::File file) : CurlDownload(url), file_(std::move(file)) {}
 };
 
+// If servers could ask the client to download any URL, there would be a security issue: the URL
+// could point to something on a private network that the server shouldn't have access to. Therefore,
+// before an HTTP download commences, the client checks for a magic file named PAKSERVER in the same
+// directory where the desired file is stored. If the file begins with the right magic string (we
+// check only a prefix to avoid any fiddly newline issues), this is interpreted as permission
+// Daemon clients to download anything from that directory.
+//
+// Note that this is the same issue addressed by the Same-Origin Policy in web browsers. In that
+// model, CORS headers are used to give scripts permission to access resources. We couldn't really
+// implement that here since there is no origin tracking of cgame binaries.
+const Str::StringRef PAKSERVER_FILE_NAME = "PAKSERVER";
+const Str::StringRef PAKSERVER_FILE_CONTENT_PREFIX = "This server is configured to host Daemon engine packages";
```

Comment:
While I'm 100% in favor of such feature, I really don't like this kind of sentence.

I think it would be better to stick to one case and to avoid whitespace (a bit like [this one before the `(C)` sutff](https://github.com/kholia/OSX-KVM/blob/fb9d9094d74a6283c8c18f94d67b3012af929655/macOS-libvirt.xml#L169)).

Also, I'm not sure that's needed to reference the Dæmon engine name. Like the DPK format, this kind of mechanism can be implemented by other engines. We also don't need such long string. The magic number can be far shorter. All we need is to reduce the probability that a file with the required name and the required content can appear unintentionally. Something like `GRANTED` sounds enough to me.

Reply:
I want the magic string to be self-documenting. As such, spaces make it easier to read and I don't see the advantage to removing them. Upper/lower case, no big deal to me. I'd be open to changing it to something that doesn't include "Daemon".

Uppercasing is one possible convention for magic filenames, to make them look distinct from the other files.

Reply:
I would prefer to leverage [RFC 5785](https://tools.ietf.org/html/rfc5785) (See for example: [eff.org/.well-known/dnt-policy.txt](https://www.eff.org/.well-known/dnt-policy.txt)) and do something like:

```
.well-known/pak-server.txt
```

It requires an extra step (creating a directory), but that would not reinvent the wheel.

Also, we can imagine a game server provider to automatically add such path as an alias (like most do with letsencrypt's `.well-known/acme-challenge` directory) without requiring customers to handle it themselves.

I prefer a simple word because of the tutorial part, the shortest and the less error prone it is, the better it is, we could tell people to do that:

```sh
mkdir -p .well-known
echo granted > .well-known/pak-server.txt
```

I worry about the uppercase because I'm already annoyed by the uppercase of `DEPS` in dpk files. When we switch from PK3 to DPK extension we had opportunity to rename it (even put it in a folder so we can easily add feature to the format), but I did not see the need for it. Now that I wrote a tool that automatically port tremulous maps by taking care of all issues including casing by lowercasing everything… I had to write a special rule in my tool to exclude README.md and… DEPS, a file we had control over. So, by default I'm a bit unpleased about uppercase files. In any way if we go the `.well-known` way it would be inconsistent to create an uppercase file name in a lowercase directory. I have nothing against uppercase content in anyway, but it's just require extra typing for nothing then it's more error prone, and there would be people to ask if it's required to be uppercase or not.

Reply:
Hmmm, `.well-known` mechanism is expected to be on `/`, not in a subdirectory, and we can serve paks in web servers subdirectories. Maybe it's a bad idea to do `.well-known` stuff the way `.well-known` stuff was never meant…

So to sum it up, my feeling is:

- it's not bad if it's a dot-hidden file (I have a strong feeling for this)
- it's not bad if there is only one string without space, even better one word (I have strong feeling for this)
- it's not bad if the case is consistent in name and in file (I have strong feeling for this)
- it's not bad if it's lower case (but I would not enforce it)

Reply:
I have some concerns about using a dot name. If there are some HTTP servers that don't serve dot files by default, that could cause headaches. Also I prefer it to show up in the HTTP server's directory listing, so that someone who is setting up a server can easily imitate a working example.

I can see the point in shortening the file contents, but one word is too vague and ungoogleable. How about `ALLOW_UNRESTRICTED_DOWNLOAD` or some case/punctuation variant thereof?

Also, I have realized an annoying flaw in the current version of the PR, which is that if you have paks in subdirectories, you'd have to create a PAKSERVER file in all subdirectories (I didn't realize it's possible to have `/` in pak names when I wrote the code). So now I'm thinking of changing the download protocol to send a base URL plus a relative one, rather than a single URL. That way PAKSERVER only has to be in the base directory.

Reply:
> How about `ALLOW_UNRESTRICTED_DOWNLOAD` or some case/punctuation variant thereof?

I'm ok with `ALLOW_UNRESTRICTED_DOWNLOAD`, I think it's better to use _same-case-plus-separator_ than mixing case.

At this point we may say _"why not `ALLOW UNRESTRICTED DOWNLOAD` with spaces?"_, but I know that some keyboard layout put a non-breaking space character on [shift+space] (which is crazy, we all agree), that's why I'm not very happy with spaces between upper case letters.

> Also, I have realized an annoying flaw in the current version of the PR, which is that if you have paks in subdirectories, you'd have to create a PAKSERVER file in all subdirectories (I didn't realize it's possible to have / in pak names when I wrote the code). So now I'm thinking of changing the download protocol to send a base URL plus a relative one, rather than a single URL. That way PAKSERVER only has to be in the base directory.

If I'm right, it's already _“base directory plus relative pak path”_ in engine, so it's better to keep that behavior all the way down.

Reply:
> If there are some HTTP servers that don't serve dot files by default, that could cause headaches. Also I prefer it to show up in the HTTP server's directory listing, so that someone who is setting up a server can easily imitate a working example.

Hmm, right, that's very strong. I just tested default apache and nginx default listing behavior and they both hide hidden dot files. I truly agree the behavior must be explicit and obvious, and for example people must be able to download the `PAKSERVER` file themselves to do it by imitation. So, let's go for `PAKSERVER` without dot and with caps.

Since `PAKSERVER` name is explicit, I still think the 'ENABLED' content would be enough. The `PAKSERVER` `ENABLED` combination looks explicit enough and self-explaining to me.